### ISSUE TITLE
bug: GitHub ratelimit unhandled

### ISSUE THREAD

**Issue by ILoveOpenSourceApplications** [ReVanced/revanced-manager#1776]
### Bug description

If alternative source is toggled on/off once, the manager goes into this unresponsive state where none of the patchable apps load, and if a patchable app is selected, none of the patches load either. And the changeling also stops working as it showing a loading circle instead.

https://github.com/ReVanced/revanced-manager/assets/117499019/75d2a702-58b3-4fb8-919c-8c619b838cdf

![Screenshot_20240322_025615_ReVanced Manager](https://github.com/ReVanced/revanced-manager/assets/117499019/e1878afa-f851-4969-b550-b25c44051d15)

PS: #1760 can be clubbed with this issue.

### Version of ReVanced Manager and version & name of app you are patching

v1.19.3

### Installation method

None

### ReVanced Manager logs

```shell
Nil
```

### Acknowledgements

- [X] This issue is not a duplicate of an existing bug report.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.
- [X] The bug is only related to ReVanced Manager

**Comment by oSumAtrIX** [comment]
Are you rate limited on GitHub? https://api.github.com/rate_limit

**Comment by ILoveOpenSourceApplications** [comment]
```
{"resources":{"core":{"limit":60,"remaining":2,"reset":1711061079,"used":58,"resource":"core"},"graphql":{"limit":0,"remaining":0,"reset":1711061595,"used":0,"resource":"graphql"},"integration_manifest":{"limit":5000,"remaining":5000,"reset":1711061595,"used":0,"resource":"integration_manifest"},"search":{"limit":10,"remaining":10,"reset":1711058055,"used":0,"resource":"search"}},"rate":{"limit":60,"remaining":2,"reset":1711061079,"used":58,"resource":"core"}}
```

**Comment by oSumAtrIX** [comment]
You seem rate-limited, but still, this is something ReVanced Manager should take care of.

**Comment by Maccer1** [comment]
I have the same issue I believe. I don't have many technical details but I am unable to patch any app and everything looks pink like in the video above.

I am using the default config and API. Tried reinstalling the last 2 versions with no luck.

**Comment by oSumAtrIX** [comment]
For support move to the links on https://revanced.app

**Comment by validcube** [comment]
👋 Hello, @ILoveOpenSourceApplications!

Quick status update here, Unfortunately I will have to unassign myself from this issue (and any other issue that I've assigned to) as tomorrow is the day where I will became unavailable for the rest of the month for unspecified reason (for team refer to recreated janky excel file in the email that I've sent about weeks ago).

🔧 For anyone who's going to work on this

From a quick glance at this issue we would need some kind of interceptor to check if GitHub reply with `429` HTTP-code or check for `x-ratelimit-remaining` HTTP header. If limit is reached we would likely inform the user about the rate-limit problem and tell them to try later (and/or inform them about the reset time: `x-ratelimit-reset` header).

Feel free to deviate from this plan if you feel like you have a better solution.
* https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit

**Comment by ILoveOpenSourceApplications** [comment]
> Feel free to deviate from this plan if you feel like you have a better solution.

Hi, @validcube. Nope, I don't have any better solution. In my exploring of the internet, I found this suggestion by someone else though. Pasting it here, if it helps, it helps.

If you're experiencing a rapid increase in API requests to GitHub and reaching the rate limit quickly, it's important to investigate what might be causing this behavior. Here are steps you can take to identify the source of the excessive API requests:

Check Application Behavior:
Start by reviewing the behavior of the third-party application that you've authorized to interact with your GitHub repositories. It's possible that this application is making a large number of requests, either unintentionally or because of a misconfiguration. Reach out to the application's support or technical team to ensure it's not making excessive requests.

Rate Limit Details: The rate limit response you posted indicates that you have a limit of 5,000 requests per hour, and it appears to be reaching that limit quickly. GitHub's rate limit includes both authenticated and unauthenticated requests. Ensure that the third-party application is using authentication (tokens) and not making unauthenticated requests, which have lower limits.

GitHub Webhooks: Check if there are any GitHub webhooks set up in your repositories. Webhooks can trigger actions in response to events in the repository, and these actions can sometimes generate API requests. Review the webhook configurations and remove any unnecessary or excessive webhooks.

GitHub Actions: If you have GitHub Actions workflows in your repositories, they could also generate API requests depending on their configuration. Inspect your workflows and make sure they are not causing excessive requests. Ensure they are configured to run only when necessary.

Check Sourcetree: Sourcetree is a Git client that interacts with your repositories, and it may use the GitHub API for certain actions. Check if Sourcetree is misconfigured or causing unexpected API requests. You can inspect its settings or documentation to understand its behavior.

Review Repository Activity: Look at your GitHub repository's "Insights" tab to see the recent activity. This can help you identify if there's any unusual or unexpected activity.

Token Scanning: GitHub has a feature called "Token Scanning" that can help you detect and revoke tokens that may be exposed or compromised. Check if any tokens have been flagged for scanning in your repository settings.

Search for Personal Access Tokens: Review the Personal Access Tokens (PATs) associated with your organization. Sometimes, developers may create tokens for various purposes, and these tokens could be the source of API requests. Revoke any tokens that are not necessary.

Rate Limit Increase: If your organization's needs require a higher rate limit, you can consider applying for an increase in your rate limits through GitHub Support. They may grant a higher limit based on your organization's usage.

Log Analysis: If possible, review server logs for the third-party application or any other services interacting with your repositories. Logs can provide insights into the frequency and nature of requests.

GitHub Audit Log: GitHub provides an audit log for organizations that can help you track actions taken by users and applications. Review the audit log to see if there are any suspicious or unexpected actions.

By systematically checking these potential sources of API requests and communication with your repositories, you should be able to identify what is causing the rapid increase in API requests and address the issue effectively. If the problem persists or you need further assistance, consider reaching out to GitHub Support for guidance and assistance with diagnosing the issue.

**Comment by ILoveOpenSourceApplications** [comment]
> Feel free to deviate from this plan if you feel like you have a better solution.

Hi, @validcube. Nope, I don't have any better solution. Though due to this issue, at times when I'm trying to patch, the manager becomes basically unresponsive and buffers and or doesn't load anything required. Switching between WiFi and mobile network seems to help at times.

**Comment by ILoveOpenSourceApplications** [comment]
This is just my doubt, but this issue began only after the introduction of changelogs into the manager. And everytime I open the manager and tap on the changelogs button, I noticed that it's fetching real time data from the internet (from the buffering). Could this also be one of the reason why the GitHub rate limit is getting exceeded as the manager now accesses the github api for fetching the changelogs for both the manager and patches as well as fetching the patch & integration data for patching as well?

If so, the making the manager remember that last changeling until an update is done to either would be better in handling the traffic right?

**Comment by kitadai31** [comment]
@validcube May I make a PR for this?

I'm thinking of using toast to notify the rate limit and time remaining like this.

<img src="https://github.com/ReVanced/revanced-manager/assets/90122968/38c4ae03-f235-4fe0-bb4d-a1032280ecb8" width="360">

But before I open the PR, I have one question.
I think the manager can show all network errors in this way.

<img src="https://github.com/ReVanced/revanced-manager/assets/90122968/06ca867c-d303-4f01-9fa2-ceb6f37dfbeb" width="360">

This is useful when the ReVanced API is down and returns 500 error, or user mistyped alternative sources.
Is this OK?

**Comment by oSumAtrIX** [comment]
Instead of a toast which can go away, you can show a label at the places where stuff should've loaded

**Comment by ILoveOpenSourceApplications** [comment]
> <img alt="" width="360" src="https://private-user-images.githubusercontent.com/90122968/331039636-38c4ae03-f235-4fe0-bb4d-a1032280ecb8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NjM1NzgsIm5iZiI6MTcxNTg2MzI3OCwicGF0aCI6Ii85MDEyMjk2OC8zMzEwMzk2MzYtMzhjNGFlMDMtZjIzNS00ZmUwLWJiNGQtYTEwMzIyODBlY2I4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEyNDExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkNDE4MjhkMGZlYzBhNjNiMzQxYjFlMWIxYTg0YzFhNjc1MWRhZWYwOGUzNzU4NTQxYjE2YjNhZGNkYTg1NDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yJD1hZbt1DGWJ1bsY0iAyJKk_I4nS0qcBX3oeh6JDFU">

Is this the best solution that's available, giving a user who wants to patch an app to wait 53 minutes for the manager to be ready? Can't we restrict something somewhere so that the manager doesn't use up the available GitHub ratelimit?

**Comment by oSumAtrIX** [comment]
A setting could be added to input a GitHub PAT to circumvent ratelimit. The ratelimit currently happens due to a bug in ReVanced Manager making more requests than necessary due a caching issue. The solution isnt known, but if found, rate limiting would hardly be a problem. You'd need to reopen ReVanced Manager like 30 times to hit any rate limits

**Comment by oSumAtrIX** [comment]
Additionally ReVanced Manager is supposed to cache everything. After a ratelimit, it should simply use the cached files, just like if offline

**Comment by validcube** [comment]
@kitadai31 

> @validcube May I make a PR for this?

Sure, you can make a PR whenever you like.

> I'm thinking of using toast to notify the rate limit and time remaining like this.
> 
> But before I open the PR, I have one question. I think the manager can show all network errors in this way.
> 
> This is useful when the ReVanced API is down and returns 500 error, or user mistyped alternative sources. Is this OK?

For me, I would use a dialog component for something critical because toast feels like simple low-priority feedback ([probably because it is](https://developer.android.com/guide/topics/ui/notifiers/toasts#show-toast:~:text=A%20toast%20provides%20simple%20feedback%20about%20an%20operation%20in%20a%20small%20popup.)) and isn't something the user would normally see or care about.

![image](https://github.com/ReVanced/revanced-manager/assets/93124920/f97a6389-99dd-4e51-a436-a6801a3492eb)
##### This is a mockup of what the dialog could look like

**Comment by ILoveOpenSourceApplications** [comment]
The current solution is a workaround rather than a fix. But atleast the user will know why nothing is loading and why they're unable to use the manager.

**Comment by oSumAtrIX** [comment]
The cache should be fixed if it is an imminent issue that needs to be looked into, then a proper solution can be done for ratelimits

**Comment by kitadai31** [comment]
OK oSumAtrIX and nullcube.
I can't think of a good idea to display the error, and I don't have that much skill with Flutter, so I stop doing this, sorry.

But I fixed the cache issue
The issue is occurred because the github_api.dart doesn't use Lock (synchronized) to call dio.get(), while the revanced_api.dart uses it
This causes GitHub API spamming

Before add synchronized:
Every time I launch the Manager, it accesses the API 3 to 7 times randomly
![2024-05-17 (5)](https://github.com/ReVanced/revanced-manager/assets/90122968/2e206583-8704-4dc4-ba0f-fa52bfe13998)

After add synchronized:
The problem is fixed
![2024-05-17 (1)](https://github.com/ReVanced/revanced-manager/assets/90122968/aa84329a-25c2-4764-a2e6-3724b62c875e)
it is fixed

I'll open a PR for this first

but API rate limit errors & network errors still have to be shown. I hope someone do this... 😕

**Comment by oSumAtrIX** [comment]
I remember adding the lock. Which commit introduced the regression?

**Comment by oSumAtrIX** [comment]
In regards to rate limits, ReVanced Manager should work just like if it were offline, until the rate limit elapses. Actions that require the API such as when the app launches for the first time or when you refresh the home feed should show a toast, or if better, a dialog explaining the rate limit

**Comment by kitadai31** [comment]
I looked around some revisions of github_api.dart, but Loke has never been added

**Comment by validcube** [comment]
👋 Hello @ILoveOpenSourceApplications!

Thanks for reporting the bugs, this is hopefully fixed in ReVanced Manager Flutter v1.21.0 and this issue will be closed.

If your problem hasn't been resolved, feels free to reopen the issue!
